docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,324 @@
+
+Additionally, drain rules can be restricted to a specific set of Machines, e.g. targeting only controlplane Machines, 
+Machines of specific Clusters, Linux Machines, etc.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: example-rule
+  namespace: default
+spec:
+  drain:
+    # behavior can be Drain or Skip
+    behavior: Drain 
+    # order can only be set if behavior == Drain
+    # Pods with higher order are drained later
+    order: 100
+  machines:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    clusterSelector:
+      matchLabels:
+      matchExpressions:
+  pods:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    namespaceSelector:
+      matchLabels:
+      matchExpressions:
+```
+
+Notes:
+
+* All selectors are of type `metav1.LabelSelector`
+* In order to allow expressing selectors with AND and OR predicates, both `machines` and `pods` have a list of 
+  selectors instead of a single selector
+
+##### Example: Exclude Pods from drain
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: skip-pods
+  namespace: default
+spec:
+  drain:
+    behavior: Skip
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchExpressions:
+      - key: app
+        operator: In
+        values:
+        - example-app1
+        - example-app2
+        - example-app3
+    namespaceSelector:
+      matchLabels:
+        kubernetes.io/metadata.name: example-namespace
+```
+
+##### Example: Drain order
+
+(related to [CAPI-11024](https://github.com/kubernetes-sigs/cluster-api/issues/11024))
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: portworx
+  namespace: default
+spec:
+  drain:
+    behavior: Drain
+    order: 100
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchLabels:
+        app: portworx
+```
+
+#### Drain annotations
+
+We propose to introduce new annotations to allow workload cluster admins/users to define drain behavior 
+for Pods. These annotations would be either added directly to Pods or indirectly via Deployments, StatefulSets, etc.
+The annotations will take precedence over `MachineDrainRules` specified in the management cluster.
+
+* `cluster.x-k8s.io/drain: skip`
+* `cluster.x-k8s.io/drain-order: <order>`
+
+#### Node drain behavior
+
+The following describes the new algorithm that decides which Pods should be drained and in which order.
+
+* List all Pods on the Node
+* For each Pod (first match applies):
+  * If the Pod is a DaemonSet Pod or a static Pod
+    * \=\> use `behavior: Skip`
+  * If the Pod has `cluster.x-k8s.io/drain: skip`
+    * \=\> use `behavior: Skip`
+  * if the Pod has `cluster.x-k8s.io/drain-order: <order>`
+    * \=\> use `behavior: Drain` and `order: <order>`
+  * If there is a matching `MachineDrainRule`
+    * \=\> use `behavior` and `order` from the first matching `MachineDrainRule`
```

Comment:
so if there's multiple conflicting rules, the first one wins? what's the order criteria?

Reply:
I mentioned it slightly below, it's alphabetical (I'll add it here as well).

It's not great but it seems like a reasonable option that doesn't overcomplicate the design and UX (one alternative would have been to add a priority field to the CRD, but then what if two CRDs have the same)

Reply:
Done